service/internal/nodecount/node_count.go

```diff
@@ -0,0 +1,154 @@
+package nodecount
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/service/controller/controllercontext"
+	"github.com/giantswarm/cluster-operator/service/internal/nodecount/internal/cache"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+}
+
+type NodeCount struct {
+	k8sClient k8sclient.Interface
+
+	nodesCache *cache.Nodes
+}
+
+func New(c Config) (*NodeCount, error) {
+	if c.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", c)
+	}
+
+	nc := &NodeCount{
+		k8sClient: c.K8sClient,
+
+		nodesCache: cache.NewNodes(),
+	}
+
+	return nc, nil
+}
+
+func (nc *NodeCount) MasterCount(ctx context.Context, obj interface{}) (map[string]Node, error) {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	nodes, err := nc.cachedNodes(ctx, cr)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	masterCount := make(map[string]Node)
+	for _, node := range nodes.Items {
+		if _, ok := node.Labels[label.MasterNodeRole]; ok {
+			id := node.Labels[label.ControlPlane]
+			{
+				val := masterCount[id]
+				val.Nodes++
+				masterCount[id] = val
+			}
+			for _, c := range node.Status.Conditions {
+				if c.Type == corev1.NodeReady && c.Status == corev1.ConditionTrue {
+					val := masterCount[id]
+					val.Ready++
+					masterCount[id] = val
+				}
+			}
+		}
+	}
+
+	return masterCount, nil
+}
+
+func (nc *NodeCount) WorkerCount(ctx context.Context, obj interface{}) (map[string]Node, error) {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	nodes, err := nc.cachedNodes(ctx, cr)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	workerCount := make(map[string]Node)
+	for _, node := range nodes.Items {
+		if _, ok := node.Labels[fmt.Sprintf("!%s", label.MasterNodeRole)]; ok {
```

Comment:
I am a bit baffled how this should work. Labels is just a string map. How do the worker role labels look like out there in the wild? What I can see in real Tenant Clusters is this. 

```
kubernetes.io/role=master
node-role.kubernetes.io/master=
node.kubernetes.io/master=
role=master
```

```
kubernetes.io/role=worker
node-role.kubernetes.io/worker=
node.kubernetes.io/worker=
role=worker
```

The code now suggests that there is a map key like `!node-role.kubernetes.io/master`, which is not how worker nodes appear to be labeled. So my expectation would have been to simply use a worker role label to find the workers.

Reply:
Changed it to worker label, I agree it is probably simpler. I didn't saw it in the first place when showing the labels on the worker.